sstables/sstables.cc

```diff
@@ -3218,14 +3219,100 @@ mutation_source sstable::as_mutation_source() {
     });
 }
 
+class sstable::filesystem_with_s3_data_storage : public filesystem_storage {
+    shared_ptr<s3::client> _client;
+    sstring _bucket;
+
+    static constexpr size_t object_name_size = 36; // stringified UUID
+    future<sstring> get_object_uuid(const sstable& sst, component_type type);
+
+public:
+    explicit filesystem_with_s3_data_storage(sstring host, sstring bucket, sstring dir)
+            : filesystem_storage(std::move(dir))
+            , _client(s3::client::make(ipv4_addr(host)))
+            , _bucket(std::move(bucket))
+    {
+    }
+
+    virtual future<file> open_component(const sstable& sst, component_type type, open_flags flags, file_open_options options, bool check_integrity) override;
+    virtual future<data_sink> make_component_sink(sstable& sst, component_type type, io_priority_class pc) override;
+    virtual future<> wipe(const sstable& sst) noexcept override;
+};
+
+future<sstring> sstable::filesystem_with_s3_data_storage::get_object_uuid(const sstable& sst, component_type type) {
+    auto f = co_await open_file_dma(sst.filename(type), open_flags::ro);
+    std::exception_ptr ex;
+    sstring uuid;
+    try {
+        auto bufptr = allocate_aligned_buffer<char>(4096, 4096);
+        size_t sz = co_await f.dma_read(0, bufptr.get(), 4096);
+        uuid = sstring(bufptr.get(), object_name_size);
+    } catch (...) {
+        ex = std::current_exception();
+    }
+
+    co_await f.close();
+    if (ex) {
+        co_await coroutine::return_exception_ptr(ex);
+    }
+
+    sstlog.trace("read {} object for {}", uuid, type);
+    co_return uuid;
+}
+
+future<file> sstable::filesystem_with_s3_data_storage::open_component(const sstable& sst, component_type type, open_flags flags, file_open_options options, bool check_integrity) {
+    auto f = co_await filesystem_storage::open_component(sst, type, flags, options, check_integrity);
+
+    if (type == component_type::Data || type == component_type::Index) {
```

Comment:
Why not extend to all the other components?

Reply:
Mainly because it needs more care around `make_component_file_writer()` that's used for other components. I'll include it in the next version